### DISCUSSION TITLE
When trying to use example file "read_db_with_helper.php" got and error message.

### DISCUSSION THREAD

**Discussion by Kachiarginis** [PHP-S7/phosphor7#1] [General]
Hi,
Maybe i'm using somthing wrong, but i use PHP8.2 and VSCode and on loading and example "read_db_with_helper.php" i got this error message : 

```
Fatal error: Uncaught Error: Call to undefined function GetLastSocketError() in 
/Users/ap/Documents/02 - CODING/03 - WEB/htdocs/phosphor7-master/src/snap_msgsock.php:747 
Stack trace: 
#0 /Users/ap/Documents/02 - CODING/03 - WEB/htdocs/phosphor7-master/src/s7_isotcp.php(501): TMsgSocket->SckConnect() 
#1 /Users/ap/Documents/02 - CODING/03 - WEB/htdocs/phosphor7-master/src/s7_peer.php(132): TIsoTcpSocket->isoConnect() 
#2 /Users/ap/Documents/02 - CODING/03 - WEB/htdocs/phosphor7-master/src/s7_micro_client.php(874): TSnap7Peer->PeerConnect() 
#3 /Users/ap/Documents/02 - CODING/03 - WEB/htdocs/phosphor7-master/src/s7_micro_client.php(866): TSnap7MicroClient->Connect() 
#4 /Users/ap/Documents/02 - CODING/03 - WEB/htdocs/phosphor7-master/examples/read_db_with_helper.php(10): TSnap7MicroClient->ConnectTo('172.16.16.10', 0, 0) 
#5 {main} thrown in /Users/ap/Documents/02 - CODING/03 - WEB/htdocs/phosphor7-master/src/snap_msgsock.php on line 747
```

Maybe it's becouse of macosx system or there is some other problem, could you "get me on the correct road" on how to use this library?`

**Comment by dixi83** [comment]
I have no experience with macos... But my first thought: You did not modify the IP in the code or the PLC's IP

Please tell me more about your setup. like:
- what PLC are you using? (including a screenshot of the HW-config)
- IP of the PLC
- share your DB your trying to read 
- is optimized block access enabled in the DB you're trying to read?
- is PUT/GET enabled on your PLC? (https://snap7.sourceforge.net/snap7_client.html#1200_1500)

**Reply by dixi83** [reply]
maybe also... tell me your goal ;)

**Comment by Kachiarginis** [comment]
- [ ] So first of all about the goal you've asked ;)

- [ ] I want to rewrite one of my Delphi (Sydney) old application that communicates with simatic PLC. This app our company using in more than 10 objects and it does the dynamic recipes load and write from plc to SCADA systems and panels. 

- [ ] Recently one company asked us to have more than one pc to configure and manage those recipes, and in this case I need server side application to do the job. So i've decided to do the web app using php (as a backend language, cause this is my favorite language of web programming;) So i found your phosphor7 repo and decided to try it. 

- [ ] I'm considering make my own Delphi communication app using snap7 that is already implamented in my app, but decided to search in google for faster and maybe more stable way and use something already made and debugged :) and found your repo! 

- [ ] So .... I will try to make a different IP, as you told me, but i think the problem is somewhere else, cause i use xdebug plugin in vscode and i see that this exeption ocoures on loading require_once libraries, but not on the connecting to given IP function, therefor i think the problem is somewhere else :) but meanwhile did not get where ;)

- [ ] Anyway i will try to send you my configuration and some print screens in vscode so you could aee some red notted warnings in some of the php files of your repo, maybe you will get some ideas what i'm doing wrong;)

- [ ] Anyway i'm really thankfull for your support and answers!

**Comment by dixi83** [comment]
I was able to reproduse your error in a bare php docker container. after installing the php `sockets` plugin it ran.

try this `docker-compose.yaml`
```
version: '3.1'
services:
    php-fpm:
        image: php:8.3-fpm
        working_dir: /application
        volumes:
            - '.:/application'
```

clone the repo in the same directory
find the id of the docker with `docker container list`
get inside the docker with `docker exec -it <id of the php docker> bash`
install the socket plugin by `docker-php-ext-install sockets`
cd to the examples directory `cd phosphor7/examples` and run `php ./read_db_with_helper.php`
if you have the correct IP/rack/slot of the PLC it will connect and try to read the first 10 bytes from DB999

this is how mine looks like:
![image](https://github.com/PHP-S7/phosphor7/assets/10693808/97149595-2bf9-4b1e-aac0-0671e0800396)

and it give me this output:
```
# php ./read_db_with_helper.php
 int(0)
----Read Data----
Data(RAW): 40b3333300580006766e
Data REAL: 5.5999999046326
Data INT:  5767174
Data INT:  423534
```

EDIT: mistake in docker-compose

**Comment by Kachiarginis** [comment]
Thank You very much for finding the cause of error. I will try to install the sockets and then see how everything
is going on. Anyway thanks a lot of finding the problem.



Audrius Piragis / Programuotojas +37065645449 / ***@***.*** ***@***.***>
Office: +370 5 2169231 / Fax: +370 5 2169231

LT-07100 Vilnius, Lietuva / Lithuania, Kaimelio g.13 

www.telemetrija.lt <applewebdata://CC1F9F5B-C9BC-47D9-842E-5363A75BB76D/www.telemetrija.lt>
